---
title: "MSA with GIMAP Gene Family"
author: "Erin Roberts"
date: "4/6/2018"
output: html_document
---

# Goal
The goal of the present exercise are as follows:
1. Analyze the sequence similarity of GIMAP sequences as identified by HMMER search
to conserved AvrRpt2-Induced Gene 1 (AIG1) (cd01852: AIG1) via: 
  -multiple sequence alignments
  -Neighbor joined phylogenetic trees
2. Use this information to determine:
  -Should these sequences be called as different genes?
  -How may these genes have propogated in the genome


# Perform multiple sequence alignment and generate phylogenetic trees

A helpful R package for creating multiple sequence alignments is MSA. 
```{r pacakges, include=TRUE}
source("http://www.bioconductor.org/biocLite.R")
biocLite("msa")
library(msa)
library(Biostrings)
system.file("tex", "texshade.sty", package="msa") #see where required LATEX file is located
install.packages("ape")
library(ape)
install.packages("seqinr")
library(seqinr)
```

## Upload sequences

```{r sequences, echo=TRUE, include=TRUE}
#load data in as a system file, first arg i any path to other directory, 2nd arg is file name, 3rd is the package
GIMAP_sequences <- readAAStringSet('/Users/erinroberts/Documents/PhD_Research/GOMEZCHIARRI2_FILES_DAILYNOTES/DAILYNOTES_DATA/apoptosis_data_pipeline/Cvir_apoptosis_manual_annotation/GIMAP/GIMAP_from_Cvir_GFF.fasta') 
head(GIMAP_sequences)

#upload file with shortened headers
GIMAP_shortened_header <- readAAStringSet('/Users/erinroberts/Documents/PhD_Research/GOMEZCHIARRI2_FILES_DAILYNOTES/DAILYNOTES_DATA/apoptosis_data_pipeline/Cvir_apoptosis_manual_annotation/GIMAP/GIMAP_from_Cvir_GFF_shortened_header.fasta')

head(GIMAP_sequences)
```

## Run MSA using clustalw default format 

The default 
```{r clustalw, echo=TRUE, include=TRUE}
GIMAP_first_alignment <- msa(GIMAP_sequences) #uses default substitution matrix

#print the alignment split over multiple blocks of sub-sequences
print(GIMAP_first_alignment, show="complete")

#re-do for file with shortened headers
GIMAP_shortened_header_alignment <- msa(GIMAP_shortened_header)
print(GIMAP_shortened_header_alignment, show="complete")

```

## Print the multiple alignment in a more attractive format

Multiple alignments can be printed using Latex TEXshade to create a PDF. 

``` {r pretty, include = TRUE, echo=TRUE}
msaPrettyPrint(GIMAP_first_alignment, output="pdf")
#can suppress adding the labels
msaPrettyPrint(GIMAP_first_alignment, output="asis", showNames = "none", showLogo="none", askForOverwrite = FALSE)

#Print out MSA with shortened headers
msaPrettyPrint(GIMAP_shortened_header_alignment, output="pdf")

#change format of pretty print
# add output="asis" and results=asis when you want to incorporate these alignments into knitted document 
msaPrettyPrint(GIMAP_shortened_header_alignment, output="pdf", file="GIMAP_shortened_header_alignment2.pdf", showConsensus = "top", consensusColors = "ColdHot", showLogo = "bottom")
```


## Create phylogenetic trees of sequence similarity and calculate distance matrix

Though these commands output a nice tree, no bootstrap values are generated.  

```{r ape, echo =TRUE, include=TRUE}
#Convert the output of the GIMAP output to aln format
GIMAP_aln <- msaConvert(GIMAP_first_alignment, type="seqinr::alignment")

# Compute a distance matrix with the dist.alignment() function from seqinr.
GIMAP_dist <- dist.alignment(GIMAP_aln, "identity")
GIMAP_dist_matrix <- as.matrix(GIMAP_dist)

#Construct tree
GIMAP_tree <- nj(GIMAP_dist)
plot(GIMAP_tree, main="GIMAP Phlyogenetic Tree of Protein Sequences")

#perform same code with shortened header file
GIMAP_shortened_aln <- msaConvert(GIMAP_shortened_header_alignment, type="seqinr::alignment")
GIMAP_shortened_dist<- dist.alignment(GIMAP_shortened_aln, "identity")
GIMAP_shortened_tree <- nj(GIMAP_shortened_dist)
plot(GIMAP_shortened_tree, main="GIMAP Phylogenetic Tree of Protein Sequences", cex=0.5)
```
toBibtex(citation("msa"))


## Create Phylogenetic Tree with Bootstrap values 

A phylip file was first generated using ClustalX gui. This will now be uploaded into R and used to generate a tree with bootstrap values. The clustalx phylip alignment was generated with default parameters. Had to convert to phylip in geneious because the phylip generated from clustalx  does not 

```{r load, echo=TRUE, include=TRUE}
library(seqinr)
GIMAP_clustalx_alignment <- read.alignment(file="/Users/erinroberts/Documents/PhD_Research/GOMEZCHIARRI2_FILES_DAILYNOTES/DAILYNOTES_DATA/apoptosis_data_pipeline/Cvir_apoptosis_manual_annotation/GIMAP/GIMAP_from_Cvir_GFF_shortest_header.phy", format="phylip") 

#the list variable now has elemtns “nb”, “nam”, “seq”, and “com”.
head(GIMAP_clustalx_alignment)
#NOTE: the names are truncated and XP was cut off and so was the last number (all wew .1 so it doesnt matter) 
```

Now we can build a rooted tree. First we set up the following function using the neighbor joining algorithm. 

```{r bootstrap, include=TRUE}
 rootedNJtree <- function(alignment, theoutgroup, type)
  {
     # load the ape and seqinR packages:
     require("ape")
     require("seqinr")
     # define a function for making a tree:
     makemytree <- function(alignmentmat, outgroup=`theoutgroup`)
     {
        alignment <- ape::as.alignment(alignmentmat)
        if      (type == "protein")
        {
           mydist <- dist.alignment(alignment)
        }
        else if (type == "DNA")
        {
           alignmentbin <- as.DNAbin(alignment)
           mydist <- dist.dna(alignmentbin)
        }
        mytree <- nj(mydist)
        mytree <- makeLabel(mytree, space="") # get rid of spaces in tip names.
        myrootedtree <- root(mytree, outgroup, r=TRUE)
        return(myrootedtree)
     }
     # infer a tree
     mymat  <- as.matrix.alignment(alignment)
     myrootedtree <- makemytree(mymat, outgroup=theoutgroup)
     # bootstrap the tree
     myboot <- boot.phylo(myrootedtree, mymat, makemytree)
     # plot the tree:
     plot.phylo(myrootedtree, type="p", cex=0.5)  # plot the rooted phylogenetic tree
     nodelabels(myboot,cex=0.7)          # plot the bootstrap values
     mytree$node.label <- myboot   # make the bootstrap values be the node labels
     return(mytree)
  }
```

Now we can plot our rooted tree after defining the outgroup. Lets set the outgroup to be GIMAP-8 which has the least number of sequences, and which were both clustered together in the previous tree without bootstrap values. 
```{r tree, echo=TRUE, include=TRUE}
GIMAP_bootstrap_tree <- rootedNJtree(GIMAP_clustalx_alignment,"022316749.", type="protein")
# this does not look pretty!
```

## Method 3 for producing trees 
Next we will use the ape package to calculate trees based on genetic distance

```{r ape, include=TRUE}
library(ape)
install.packages("adegenet")
library(adegenet)
library(stats)
GIMAP_mRNA <- fasta2DNAbin(file="/Users/erinroberts/Documents/PhD_Research/GOMEZCHIARRI2_FILES_DAILYNOTES/DAILYNOTES_DATA/apoptosis_data_pipeline/Cvir_apoptosis_manual_annotation/GIMAP/GIMAP_from_Cvir_GFF_shortest_header.fasta", chunks=10)
GIMAP_dist <- dist.dna(GIMAP_mRNA, model="TN93")
```

## Calculate Consensus Sequences and Conservation Scores

```{r consensus, include=TRUE}
printSplitString <- function(x, width=getOption("width") - 1)
f
starts <- seq(from=1, to=nchar(x), by=width)
for (i in 1:length(starts))
cat(substr(x, starts[i], starts[i] + width - 1), "nn")
g
printSplitString(msaConsensusSequence(GIMAP_shortened_header_alignment))
```


## Helpful websites for reference
http://a-little-book-of-r-for-bioinformatics.readthedocs.io/en/latest/src/chapter5.html
MSA documentation 
http://adegenet.r-forge.r-project.org/files/PRstats/practical-introphylo.1.0.pdf