---
title: "GIMAP IAP Haplotig Search Report"
author: "Erin Roberts"
date: "6/9/2020"
output: pdf_document
classoption: landscape
---

```{r, include=FALSE}
options(tinytex.verbose = TRUE)
```

Goal: Investigate presence of haplotigs in the genome annotation of *C. virginica* IAP and GIMAP gene families. 

## **Methods**
1. Compare sequence identity of all protein hits using CD-Hit at a level of 95% sequence identity
    - "global sequence identity" calculated as: number of identical amino acids or bases in alignment, divided by the full length of the shorter sequence
    - Code used in bluewaves:
      `module load CD-HIT/4.8.1-foss-2018b
cd-hit -G 1 -c 1.0 -t 1 -i $F/BIR_IAP_HMMER_Interpro_XP_list_all.fa -o $F/BIR_IAP_HMMER_Interpro_XP_list_all_rm_dup.fa
echo "done rm dup $(date)`
2. Merge CD-Hit protein sequence identity data with Jon Puritz data of mean coverage across all individuals.
3. Compare with Haplotig finder output (need to get software info from Jon) of genomic locations with haplotigs.
4. Identify protein clusters containing sequences from two different genes.
5. Compute average coverage within clusters, and then comparing between clusters.
5. Identify protein clusters with average coverage < 500 (about half of the coverage across most clusters).
6. Align nucleotide sequences from clusters and view alignments with CDS and haplotig information. 

```{r lib, include=FALSE, echo=FALSE}
library(ape)
library(Biostrings)
library(ggplot2)
library(ggtree) # install the dev version to get the get.tree function
library(ggrepel)
library(phylotools)
library(treeio)
library(tidytree)
library(ggimage)
library(plyr)
library(tidyverse)
#library(tidytext)
library(rtracklayer)
library(data.table)
library(chopper)
library(alakazam)
library(phylotools)

```

### Load BED files from Jon 
JP provided BED files for each of the two gene families, with mean coverage values averaged across all 90 individuals. Each gene has a coverage value (see below).

```{r loadBED, include=TRUE}
Cvir_GIMAP_meanCov <- read.table(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Jon_Puritz_6_4_2020_regenefamilybedfiles/Cvir_GIMAP.meanCov.bed",
                                 sep="\t", col.names = c("seqid","start","end","gene","meanCov"))
head(Cvir_GIMAP_meanCov)
```


```{r loadBED2, include=TRUE}
Cvir_IAP_meanCov <- read.table(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Jon_Puritz_6_4_2020_regenefamilybedfiles/Cvir_IAP.meanCov.bed",
                                 sep="\t", col.names = c("seqid","start","end","gene","meanCov"))
```
The data format for haplotigs file lists large regions in the genome where haplotigs were identified. All the counts are 0 (not sure what this means). Each identified haplotig encompasses many genes, not just a single gene.


```{r loadBED3, include=TRUE}
Cvir_haplotigs <- read.table(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Jon_Puritz_6_4_2020_regenefamilybedfiles/haplotigs.bed",
                               sep="\t", skip= 1, col.names = c("seqid","start","end","counts","dataset"))
head(Cvir_haplotigs)
```

```{r load, echo=FALSE}
# Load data generated in longer `IAP_GIMAP_Gene_Family_Expansion.R` script 

load("/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/AIG_seq_rm_dup_clstr6_dup_diff_gene_95_product_95.Rdata")
load("/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/BIR_seq_rm_dup_clstr6_dup_diff_gene_product_95.Rdata")

load(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Cvir_IAP_EMR_Name.Rdata")
load(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Cvir_GIMAP_EMR_Name.Rdata")

load(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/AIG_seq_rm_dup_clstr6_NUC_95.Rdata")
load(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/BIR_seq_rm_dup_clstr6_NUC_95.Rdata")

```    

### Investigate CD-hit results
Let's now review the results from CD-Hit and join the mean coverage information. CD-Hit software works by first clustering sequences by sequence similarity. Proteins in the cluster denoted by a `*` are the longest sequence in the cluster and are used as the reference sequence in the cluster. Sequences in the cluster denoted with a similarity percentage are that percentage identical to the sequence denoted with `*`. 

#### Join the CD-Hit results for each family with mean coverage
In order to narrow down the CD-Hit clusters to view, we are only investigating the clusters where proteins were clustered across two different genes.  

```{r mergeBED, include=TRUE, message=FALSE}
Cvir_GIMAP_meanCov_CD_Hit_95 <- left_join(AIG_seq_rm_dup_clstr6_dup_diff_gene_95_product_95 , 
                 Cvir_GIMAP_meanCov) %>% filter(Species =="Crassostrea_virginica")
Cvir_IAP_meanCov_CD_Hit_95 <- left_join(BIR_seq_rm_dup_clstr6_dup_diff_gene_product_95 , 
                  Cvir_IAP_meanCov) %>% filter(Species =="Crassostrea_virginica")
```

### Join Gene Length with CD-Hit and Mean Coverage results 
Ximing pointed out that haplotigs are likely 1 or more MB in length and are large sequences. Joining the nucleotide length of each to see if this is informative.  

```{r genelength, include=TRUE, message=FALSE}
Cvir_GIMAP_meanCov_CD_Hit_95 <- left_join(Cvir_GIMAP_meanCov_CD_Hit_95, GIMAP_BED_name)
Cvir_IAP_meanCov_CD_Hit_95 <- left_join(Cvir_IAP_meanCov_CD_Hit_95, IAP_BED_name)

GIMAP_gene_length_aa <- AIG_seq_rm_dup_clstr6_NUC_95[,c("aa","gene")]
IAP_gene_length_aa   <- BIR_seq_rm_dup_clstr6_NUC_95[,c("aa","gene")]
colnames(GIMAP_gene_length_aa)[1] <- "gene_length"
colnames(IAP_gene_length_aa  )[1] <- "gene_length"

Cvir_GIMAP_meanCov_CD_Hit_95_length <- left_join(Cvir_GIMAP_meanCov_CD_Hit_95, GIMAP_gene_length_aa)
Cvir_IAP_meanCov_CD_Hit_95_length <- left_join(Cvir_IAP_meanCov_CD_Hit_95 , IAP_gene_length_aa)
Cvir_IAP_meanCov_CD_Hit_95_length <- unique(Cvir_IAP_meanCov_CD_Hit_95_length)

# Make unique for each gene
Cvir_GIMAP_meanCov_CD_Hit_95_length_unique <- Cvir_GIMAP_meanCov_CD_Hit_95_length %>% distinct(gene, .keep_all = TRUE)
Cvir_IAP_meanCov_CD_Hit_95_length_unique <-  Cvir_IAP_meanCov_CD_Hit_95_length %>% distinct(gene, .keep_all = TRUE)
```

\newpage
### View GIMAP combined CD-Hit, gene length, mean coverage results
Lets first view the GIMAP gene family results. 

```{r genelengthGIMAP, echo=FALSE, R.options=list(width=120)}
print(Cvir_GIMAP_meanCov_CD_Hit_95_length_unique[,-c(2,3,7:11)], row.names=F)
```

**Do any clusters with two genes with high protein simarilty have a large difference in coverage?** 

1. Cluster 65: genes LOC111110237, LOC111110097 have the largest difference in coverage within clusters
2. Cluster 219: gene LOC111110115, LOC111106081 genes have the lowest coverage compared to genes in other clusters

### Review Gene Location on GIMAP tree
```{r gene_biom_down, echo=FALSE, message=FALSE, warning=FALSE, fig.height= 10 , fig.width=10}
load(file="/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/GIMAP_raxml_treedata.Rdata")

GIMAP_raxml_treedata_pomacea_down_subset <- tree_subset(GIMAP_raxml_treedata, 60, levels_back = 7)
 ggtree(GIMAP_raxml_treedata_pomacea_down_subset, layout="circular", aes(color=Species), branch.length = "none") + 
  geom_tiplab2(aes(label=gene_locus_tag,angle=angle), size =2.2, offset=.5) + # geom_tiplab2 flips the labels correctly
  theme(legend.position = "right", legend.text = element_text(face = "italic")) + xlim(-70,70)  

```

\newpage

### View IAP combined CD-Hit, gene length, mean coverage results
Lets view the IAP gene family results. 

```{r genelengthIAP, echo=FALSE, R.options=list(width=130)}
print(Cvir_IAP_meanCov_CD_Hit_95_length_unique[,-c(2,3,7:11)], row.names=F)
```


**Do any clusters with two genes with high protein simarilty have a large difference in coverage?** 

  - Cluster 17: LOC111100400 or LOC111100443 Cluster 17 has two genes with large differences in coverage. One has about half coverage and LOC111100400 is very long
  - Cluster 280: LOC111103392 has about half coverage compared to other gene
  - Cluster 282: LOC111104229 has half coverage as compared to the other gene in the cluster and very high sequence similarity
  - cluster 328: Coverage is variable throughout cluster. 
  - cluster 338: Two have high coverage around 1500 and LOC111116378 has 189 coverage
  - Cluster 344:LOC111111659 and LOC111116826 both have very low coverage compared to LOC111117856

RAxML tree is still run on the cluster so we can't review tree location.


### Compare with Haplomerger results 
Lets now investigate if these interesting gene clusters overlap with regions identified as haplotigs using JP software. Because the haplotig tool identified large regions, I am going to search for whether my cluster genes of interest within these regions. If both the gene start and gene end of my genes are inside any haplotig tool identified range, both of these columns are denoted with a "YES".  

```{r haplorange, echo=TRUE}
Cvir_GIMAP_meanCov_CD_Hit_95_length_unique$HM_found_start <- ifelse(sapply(Cvir_GIMAP_meanCov_CD_Hit_95_length_unique$start, function(p) 
  any(Cvir_haplotigs$start <= p & Cvir_haplotigs$end >= p)),"YES", NA)
Cvir_GIMAP_meanCov_CD_Hit_95_length_unique$HM_found_end <- ifelse(sapply(Cvir_GIMAP_meanCov_CD_Hit_95_length_unique$end, function(p) 
  any(Cvir_haplotigs$start <= p & Cvir_haplotigs$end >= p)),"YES", NA)

Cvir_IAP_meanCov_CD_Hit_95_length_unique$HM_found_start <- ifelse(sapply(Cvir_IAP_meanCov_CD_Hit_95_length_unique$start, function(p) 
  any(Cvir_haplotigs$start <= p & Cvir_haplotigs$end >= p)),"YES", NA)
Cvir_IAP_meanCov_CD_Hit_95_length_unique$HM_found_end <- ifelse(sapply(Cvir_IAP_meanCov_CD_Hit_95_length_unique$end, function(p) 
  any(Cvir_haplotigs$start <= p & Cvir_haplotigs$end >= p)),"YES", NA)
```

### View GIMAP cluster overlap with haplotig results
"YES" in both columns indicates that these genes are indeed a region identified as a haplotig, while "NA" means the genes are not inside a region identified via software to contain haplotigs.

```{r haplorangeGIMAP, echo=FALSE,  R.options=list(width=130)}
print(Cvir_GIMAP_meanCov_CD_Hit_95_length_unique[,-c(2,3,7:11)], row.names=F)
```

\newpage
### View IAP cluster overlap with haplotig results

```{r haplorangeIAP, echo=FALSE,  R.options=list(width=130)}
print(Cvir_IAP_meanCov_CD_Hit_95_length_unique[,-c(2,3,7:11)], row.names=F)
```

Overall, for both gene families, most genes identified in these clusters are inside ranges called as haplotigs.


### Questions and Observations:
1. What is the expected coverage for each gene? We estimated from these results the average coverage for genes within clusters is around 1000. 
2. What is the best way to identify clusters containing haplotigs? 
    - Two potential methods:
    - 1) Investigate clusters where one gene has "normal" coverage and the other gene has half normal (what I did above in my results for each family).
    - 2) Take the average of mean coverage and investigate clusters where the average gene coverage across the cluster is about half of what we would expect (<500)

3. Is the haplotig finding tool over-assigning haplotigs?    

Moving forward we decided to take approach 2.2 above to narrow clusters to investigate, since clusters identified with strategy 1.2 may just caused by rare genes across populations. 

# Average the mean gene coverage within clusters 

### Calculate mean coverage within gene clusters
```{r meanCov_clstr, include=TRUE}
Cvir_GIMAP_meanCov_CD_Hit_95_length_unique_mean <- Cvir_GIMAP_meanCov_CD_Hit_95_length_unique %>% group_by(cluster) %>%
  mutate(mean_Cov_clstr = mean(meanCov))
Cvir_IAP_meanCov_CD_Hit_95_length_unique_mean <- Cvir_IAP_meanCov_CD_Hit_95_length_unique %>% group_by(cluster) %>%
  mutate(mean_Cov_clstr = mean(meanCov)) 
```

### View GIMAP mean coverage results
```{r meanCov_clstrGIMAP, echo=FALSE, R.options=list(width=130)}
print(Cvir_GIMAP_meanCov_CD_Hit_95_length_unique_mean[,-c(2,3,7:11)], row.names=F)
```

**GIMAP results**:

- Cluster 291: Mean coverage of 436. Inspected the nucleotide sequence alignment of these genes LOC111110115, LOC111106081 below.

\newpage
### View IAP mean coverage results
```{r meanCov_clstrIAP, echo=FALSE, R.options=list(width=130)}
print(Cvir_IAP_meanCov_CD_Hit_95_length_unique_mean[,-c(2,3,7:11)], row.names=F)
```

**IAP results:**

- Cluster 62: mean coverage of 444. Includes LOC111100470 and LOC111101689
- Cluster 328: mean coverage across cluster of 280. Includes LOC111132301 LOC111114013, LOC111103682, LOC111132489, LOC111132589, LOC111102106, LOC111114070
- Cluster 344: mean coverage across cluster 484. Includes LOC111117856, LOC111116826, LOC111111659

# Inspect Gene Clusters 
Now let's investigate these clusters more closely. The genes identified in each cluster were aligned using MAFFT with default settings and visualized in Unipro UGENE. 

```{r cluster, include=FALSE}
GIMAP_cluster_219_gene <- Cvir_GIMAP_meanCov_CD_Hit_95_length_unique_mean  %>% filter(cluster == "Cluster 219")
IAP_cluster_62_gene <- Cvir_IAP_meanCov_CD_Hit_95_length_unique_mean  %>% filter(cluster == "Cluster 62") 
IAP_cluster_328_gene <- Cvir_IAP_meanCov_CD_Hit_95_length_unique_mean  %>% filter(cluster == "Cluster 328")
IAP_cluster_344_gene <- Cvir_IAP_meanCov_CD_Hit_95_length_unique_mean  %>% filter(cluster == "Cluster 344") 
```

### GIMAP Cluster 219

Zooming in on GIMAP cluster 219
```{r 219, echo=FALSE, R.options=list(width=130)}
print(GIMAP_cluster_219_gene[,c("seqid","start","end","prot_identity_stat","cluster", "meanCov","mean_Cov_clstr")])
```

Both sequences are on the same chromosome. Lets take a look now at several sections of nucleotide alignment of these two genes where there are sequence insertions. 

![GIMAP 219 section 1 insertion alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/GIMAP_219_section1.png)

Overall most of the nucleotide sequences have complete identity, except for several sections with insertions or deletions. 

![GIMAP 219 section 2 insertion alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/GIMAP_219_section2.png)

![GIMAP 219 section 3 insertion alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/GIMAP_219_section3.png)

![GIMAP 219 section 4.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/GIMAP_219_section4.png)

**Cluster 219 Conclusion:**
The two genes in the GIMAP protein cluster 219 should be collapsed into one based on high sequence identity and low coverage compared to other clusters where genes are similar.

### IAP Cluster 62 

Zooming in on IAP cluster 62
```{r 62, echo=FALSE, R.options=list(width=130)}
print(IAP_cluster_62_gene[,c("seqid","start","end","prot_identity_stat","cluster", "meanCov","mean_Cov_clstr")])
```

![IAP cluster 62 alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/IAP_cluster_62.png)

**Cluster 62 Conclusion:**
The two genes in the IAP protein cluster 62 should not be collapsed because they have low nucleotide level sequence identity and likely are truly different genes despite having very low coverage.

### IAP Cluster 328
Zooming in on IAP cluster 328
```{r 328, echo=FALSE, R.options=list(width=130)}
print(IAP_cluster_328_gene[,c("seqid","gene","prot_identity_stat","cluster", "meanCov","mean_Cov_clstr")])
```

These 7 genes found on different chromosomes have an odd distribution of coverage, with two genes appearing to be "real" beacuse of higher relative coverage, and the other 5 appearing to be artifacts. 

![IAP cluster 328 section 1 alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/IAP_cluster_328_section1.png)

![IAP cluster 328 section 2 alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/IAP_cluster_328_section2.png)

![IAP cluster 328 section 3 alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/IAP_cluster_328_section3.png)

**Cluster 328 Conclusion:**
The two genes on top of the alignment (LOC111132489 and LOC111114013) form a cluster together and are most similar and should be collapsed into one gene. One has an individual coverage of 890 while the other has coverage of 103. The five other genes are very similar in nucleotide sequence and all have relatively low coverage, though LOC111132301 has the highest relatively. These five genes should also be collapsed together. 

### IAP Cluster 344
Zooming in on IAP cluster 344
```{r 344, echo=FALSE, R.options=list(width=130)}
print(IAP_cluster_344_gene[,c("seqid","start","end","gene","prot_identity_stat","cluster", "meanCov","mean_Cov_clstr")])
```

![IAP cluster 344 section 1 alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/IAP_cluster_344_section1.png)

![IAP cluster 344 section 2 alignment.](/Users/erinroberts/Documents/PhD_Research/Chapter_1_Apoptosis Paper/Chapter_1_Apoptosis_Annotation_Data_Analyses_2019/DATA/Apoptosis_Pathway_Annotation_Comparative_Genomics/Comparative_Analysis_Apoptosis_Gene_Families_Data/Gene_Artifact_Investigation/IAP_cluster_344_section2.png)

Notice that in this cluster, two genes on the same chromosome, while one is on separate chromosome. The two sequences with the greatest similarity in gene sequence are LOC111116826 and LOC111111659. These two genes should be collapsed into 1. 

## Final Additional Conclusions

1. If the methods used here to confirm haplotigs seem valid, the haplotig identifying tool may be over assigning haplotigs. 


